### PR TITLE
Adding beforeAttach and beforeDetach event handlers including tests

### PR DESCRIPTION
**Pull request by rowanwins**
This is a followup pull request on #2526 which aims to implement the beforeAttach and beforeDetach features discussed in #1469 .

Let me know if anything further is required.

### CODE REVIEW

**Comment by yyx990803** [comment]
Looking good! Will take one last pass later today to make sure everything is covered :+1:

File: test/unit/specs/api/lifecycle_spec.js [hunk 1]
```diff
@@ -112,7 +112,7 @@ describe('Lifecycle API', function () {
     })
 
     it('hooks', function () {
-      var hooks = ['created', 'beforeCompile', 'compiled', 'attached', 'ready']
+      var hooks = ['created', 'beforeCompile', 'compiled', 'beforeAttach', 'attached', 'ready']
```
Comment (vuchl) [inline test/unit/specs/api/lifecycle_spec.js]: Is `beforeDettach` missing here?

**Comment by posva** [comment]
Can you please add to the commit description `Fix #1469` so it also closes the issue?

File: src/fragment/fragment.js [hunk 2]
```diff
@@ -138,6 +145,7 @@ function multiRemove () {
   var self = this
   var shouldCallRemove = inDoc(this.node)
   this.beforeRemove()
+  self.callHook(beforeDetach)
```
Comment (yyx990803) [inline src/fragment/fragment.js]: should check `shouldCallRemove` here too

File: src/fragment/fragment.js [hunk 3]
```diff
@@ -99,6 +102,7 @@ function singleRemove () {
   this.inserted = false
   var shouldCallRemove = inDoc(this.node)
   var self = this
+  self.callHook(beforeDetach)
```
Comment (yyx990803) [inline src/fragment/fragment.js]: need to check `shouldCallRemove`

File: test/unit/specs/api/lifecycle_spec.js (hunk 1 above)
Comment (rowanwins) [inline test/unit/specs/api/lifecycle_spec.js]: @vuchl it looked like this series of tests was for monitoring adding stuff to the DOM, rather the the removal aspect but I could be wrong...

**Comment by rowanwins** [comment]
Gday @yyx990803 just checking that you saw I updated this as per your suggestion, hopefully it's close now :)

**Comment by yyx990803** [comment]
@rowanwins it is - it's just this PR has a bit more moving pieces than others so I want to be extra careful. It will probably be part of 1.1 instead of the next patch release, so please be patient :)

**Comment by rowanwins** [comment]
Gday @yyx990803 , so Im thinking this pull request may not be required if the [Vue 2.0 write up](https://github.com/vuejs/vue/wiki/2.0-features) is anything to go by... Anyway no biggie if that's the case, was good to poke around in the source and I'm looking forward to seeing what v2 brings :)

Cheers
Rowan

**Comment by yyx990803** [comment]
Hey @rowanwins - apologies for letting this PR hang for so long.

In v2, `attached` and `detached` are indeed no longer supported. This is mainly due to their usefulness not justifying the amount of complexity in ensuring correct behavior for these hooks. Given that, `beforeAttach` and `beforeDetach` obviously are not needed as well.

For v1, it would be less than ideal to introduce two hooks that are soon to be gone in v2, so unfortunately that means we will probably have to let this PR go. I know you've put a lot of work into it, so I do want to thank you for that and say sorry ;'(

**Pull request closed without merging**